src/main/scala/org/scalameter/japi/JBench.scala

```diff
@@ -0,0 +1,271 @@
+    object BenchmarkExtractor {
+      def unapply(m: Method): Option[Seq[String]] = {
+        Option(m.getAnnotation(classOf[benchmark])).map(_.value().split('.').toSeq)
+      }
+    }
+
+    val clazz = this.getClass
+    val scopedSetups: Seq[(Seq[String], Seq[Method])] = {
+      val mapping = mutable.Map.empty[Seq[String], Seq[Method]]
+      for (m @ BenchmarkExtractor(scopes) <- clazz.getMethods) {
+        mapping.get(scopes) match {
+          case Some(setups) =>
+            mapping += (scopes -> (setups :+ m))
+          case None =>
+            mapping += (scopes -> Seq(m))
+        }
+      }
+      mapping.toList
+    }
+    val scopeCtxs = {
+      Option(clazz.getAnnotation(classOf[scopes])).map { a =>
+        a.value().map { sc =>
+          sc.scope().split('.').toSeq -> getField[Context](clazz, sc.context())
+        }.toMap
+      }.getOrElse(Map.empty)
+    }
+    for ((scope, mapping) <- scopedSetups.groupBy(_ ._1.head)) {
+      setScope(clazz, scopeCtxs, scope, mapping.map(kv => (kv._1.tail, kv._2)))
+    }
+  }
+
+  /** Sets given scope.
+   *
+   *  It sets declared scopes recursively,
+   *  from top to the bottom where setup items are created.
+   */
+  private def setScope(clazz: Class[_],
+    scopeCtxs: Map[Seq[String], Context], name: String,
+    scopedSetups: Seq[(Seq[String], Seq[Method])]): Unit = {
+    val scope = Scope(name, setupzipper.value.current.context)
+    scope config {
+      scopeCtxs.getOrElse(
+        (scope.name :: scope.context(Key.dsl.scope)).reverse, Context.empty
+      )
+    } in {
+      scopedSetups.groupBy(_._1.headOption).foreach {
+        case (None, newMapping) =>
+          newMapping.flatMap(_._2).foreach(m => setSetup(clazz, m))
+        case (Some(newScope), newMapping) =>
+          setScope(
+            clazz, scopeCtxs, newScope, newMapping.map(kv => (kv._1.tail, kv._2))
+          )
+      }
+    }
+  }
+
+  /** Sets setup for a benchmark snippet.
+   */
+  private def setSetup(cl: Class[_], m: Method) {
+    val additionalContext = Option(m.getAnnotation(classOf[ctx]))
+      .map(a => getField[Context](cl, a.value()))
+      .getOrElse(Context.empty)
+
+    val gen = Option(m.getAnnotation(classOf[gen]))
+      .map(a => getField[Gen[AnyRef]](cl, a.value()))
+      .getOrElse(sys.error("Each benchmark method should be annotated with 'gen'."))
+
+    val setupBeforeAll= getNoArgMethod(m, classOf[setupBeforeAll], cl)
+      .asInstanceOf[Option[() => Unit]]
+
+    val teardownAfterAll = getNoArgMethod(m, classOf[teardownAfterAll], cl)
+      .asInstanceOf[Option[() => Unit]]
+
+    val setup: Option[AnyRef => Any] = getOneArgMethod(m, classOf[setup], cl)
+
+    val teardown: Option[AnyRef => Any] = getOneArgMethod(m, classOf[teardown], cl)
+
+    val warmup: Option[() => Any] = getNoArgMethod(m, classOf[warmup], cl)
+
+    val snippet: AnyRef => Any = {
+      m.setAccessible(true)
+      val sm = new SerializableMethod(m)
+      v => sm.invokeA(this, v)
```

Comment:
May be a good idea to do a quick sanity check - you don't need to do this for this PR, but would be useful eventually.